### PR TITLE
Support for Open3DGC-compression in glTF importer/exporter.

### PR DESCRIPTION
**Pull request by smalcom**
Extension has some limitations due unfinished specification of glTF. Also if you plan to open models generated by COLLADA2GLTF then i have bad news for you. Proposed implementation use "buffer" against "bufferView" and do not use "floatAttributes" (as unneeded).
And some information from source code:

> Limitations for meshes when using Open3DGC-compression.
> It's a current limitation of sp... Specification have not this part still - about mesh compression. Why only one primitive? Because glTF is very flexibly. But in fact it ugly flexible. Every primitive can has own set of accessors and accessors can point to a-a-a-a-any part of buffer (thru bufferview ofcourse) and even to another buffer. We know that "Open3DGC-compression" is applicable only to part of buffer. As we can't guaranty continuity of the data for decoder, we will limit quantity of primitives. Yes indices, coordinates etc. still can be stored in different buffers, but with current specification it's a exporter problem.
> Also primitive can has only one of "POSITION", "NORMAL" and less then "AI_MAX_NUMBER_OF_TEXTURECOORDS" of "TEXCOORD". All accessor of primitive must point to one continuous region of the buffer.

### CODE REVIEW

**Comment by smalcom** [comment]
Hm. I thought whole project already uses C++11. It is necessary to use older C++?

**Comment by smalcom** [comment]
Also, i'm preparing for commit importers of AMF and X3D formats.

**Comment by smalcom** [comment]
> Hm. I thought whole project already uses C++11. It is necessary to use older C++?

For example:
3dsloader.cpp
`for (auto &mesh : mScene->mMeshes)`

aseloader.cpp
`for (BaseNode *node : nodes) {`

**Comment by acgessler** [comment]
We used to be pre C++11. However, as I understand, the requirement is no longer. Keep the public API free of C++11 constructions, though.

Can you look into the failing builds?

**Comment by smalcom** [comment]
> Can you look into the failing builds?
in progress

**Comment by smalcom** [comment]
Build is checked on:
1. WinXP, mingw32-4.8.5
2. Linux, gcc-5.3.0

**Comment by smalcom** [comment]
Two errors:
1. std::min, std::max not found. I don't know why: "#include <algorithm>" is present.  Code not mine.
2. 
> CMake Error at test/CMakeLists.txt:105 (SET_PROPERTY):
> set_property could not find TARGET assimp. Perhaps it has not yet been
> created.

Also strange error, because target added in "code/CMakeLists.txt" which included to "CMakeLists.txt" at
line 324. Where directory "test" included only at line 371.

**Comment by smalcom** [comment]
Also, try to build with MSVC 2010 is failed. Because
> for(aiMesh* m : meshes_linear) {

in XGLLoader.h. Similar error from BlobIOSystem.h and lot of other files has errors.
This not one type of errors, other errors are present too.

**Comment by smalcom** [comment]
fixed

**Comment by smalcom** [comment]
something required from me for accepting pull-request? also look at
https://github.com/assimp/assimp/issues/981

**Comment by kimkulling** [comment]
I just have to do the review,   but I didn't find some time, sorry.

smalcom <notifications@github.com> schrieb am Mi., 17. Aug. 2016, 19:09:

**Comment by smalcom** [comment]
Understood. Thanks for answer.

**Comment by kimkulling** [comment]
First finding: windows build is broken:
"C:\projects\assimp\Assimp.sln" (default target) (1) ->
       "C:\projects\assimp\code\assimp.vcxproj.metaproj" (default target) (3) ->
       "C:\projects\assimp\code\assimp.vcxproj" (default target) (12) ->
       (ClCompile target) -> 
         c:\projects\assimp\code\glTFAsset.inl(509): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\assimp\code\glTFImporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(509): error C2059: syntax error: '::' (compiling source file C:\projects\assimp\code\glTFImporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1206): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\assimp\code\glTFImporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1206): error C2059: syntax error: '::' (compiling source file C:\projects\assimp\code\glTFImporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1261): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\assimp\code\glTFImporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1261): error C2059: syntax error: '::' (compiling source file C:\projects\assimp\code\glTFImporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(509): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\assimp\code\glTFExporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(509): error C2059: syntax error: '::' (compiling source file C:\projects\assimp\code\glTFExporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1206): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\assimp\code\glTFExporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1206): error C2059: syntax error: '::' (compiling source file C:\projects\assimp\code\glTFExporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1261): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\assimp\code\glTFExporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
         c:\projects\assimp\code\glTFAsset.inl(1261): error C2059: syntax error: '::' (compiling source file C:\projects\assimp\code\glTFExporter.cpp) [C:\projects\assimp\code\assimp.vcxproj]
    94 Warning(s)
    12 Error(s)

Check https://ci.appveyor.com/project/kimkulling/assimp/build/1.0.627 to get more information.

File: code/glTFAsset.h [hunk 1]
```diff
@@ -451,32 +450,115 @@ namespace glTF
...
+			{}
+
+			/// \fn ~SEncodedRegion()
+			/// Destructor.
+			~SEncodedRegion() { delete [] DecodedData; }
+		};
+
+		/******************* Variables *******************/
+
+		//std::string uri; //!< The uri of the buffer. Can be a filepath, a data uri, etc. (required)
```
Comment (kimkulling) [inline code/glTFAsset.h]: Dead code.

File: code/glTFAsset.h [hunk 2]
```diff
@@ -451,32 +450,115 @@ namespace glTF
...
+		size_t byteLength; //!< The length of the buffer in bytes. (default: 0)
+		//std::string type; //!< XMLHttpRequest responseType (default: "arraybuffer")
```
Comment (kimkulling) [inline code/glTFAsset.h]: Dead code

File: code/glTFAsset.h [hunk 3]
```diff
@@ -631,10 +710,77 @@ namespace glTF
...
+			{
+				using SExtension::Type;
+
+				std::string Buffer;///< ID of "buffer" used for storing compressed data.
+				size_t Offset;///< Offset in "bufferView" where compressed data are stored.
+				size_t Count;///< Count of elements in compressed data. Is always equivalent to size in bytes: look comments for "Type" and "Component_Type".
+				bool Binary;///< If true then "binary" mode is used for coding, if false - "ascii" mode.
+				size_t IndicesCount;///< Count of indices in mesh.
+				size_t VerticesCount;///< Count of vertices in mesh.
+				// AttribType::Value Type;///< Is always "SCALAR".
```
Comment (kimkulling) [inline code/glTFAsset.h]: Dead code

File: code/glTFAsset.h [hunk 4]
```diff
@@ -631,10 +710,77 @@ namespace glTF
...
+				// ComponentType Component_Type;///< Is always "ComponentType_UNSIGNED_BYTE" (5121).
```
Comment (kimkulling) [inline code/glTFAsset.h]: Dead code

File: code/glTFAsset.inl [hunk 5]
```diff
@@ -341,6 +427,9 @@ inline void Buffer::Grow(size_t amount)
     byteLength += amount;
 }
 
+//
+// struct BufferView
+//
```
Comment (kimkulling) [inline code/glTFAsset.inl]: Dead code

File: code/glTFAsset.inl [hunk 6]
```diff
@@ -393,7 +484,17 @@ inline uint8_t* Accessor::GetPointer()
...
     size_t offset = byteOffset + bufferView->byteOffset;
-    return basePtr + offset;
+
+	// Check if region is encoded.
+	if(bufferView->buffer->EncodedRegion_Current != nullptr)
+	{
+		const size_t begin = bufferView->buffer->EncodedRegion_Current->Offset;
+		const size_t end = bufferView->buffer->EncodedRegion_Current->Offset + bufferView->buffer->EncodedRegion_Current->DecodedData_Length;
+
+		if((offset >= begin) && (offset < end)) return &bufferView->buffer->EncodedRegion_Current->DecodedData[offset - begin];
```
Comment (kimkulling) [inline code/glTFAsset.inl]: Hard to read, maybe 2 lines?

**Comment by smalcom** [comment]
> Dead code

in which meaning?

File: code/CMakeLists.txt [hunk 7]
```diff
@@ -701,7 +701,34 @@ SET ( openddl_parser_SRCS
 )
 SOURCE_GROUP( openddl_parser FILES ${openddl_parser_SRCS})
 
+SET ( open3dgc_SRCS
+  ../contrib/Open3DGC/o3dgcArithmeticCodec.cpp
+  ../contrib/Open3DGC/o3dgcDynamicVectorDecoder.cpp
+  ../contrib/Open3DGC/o3dgcDynamicVectorEncoder.cpp
+  ../contrib/Open3DGC/o3dgcTools.cpp
+  ../contrib/Open3DGC/o3dgcTriangleFans.cpp
+  ../contrib/Open3DGC/o3dgcTimer.h
```
Comment (kimkulling) [inline code/CMakeLists.txt]: The common file header is missing, right?

**Comment by kimkulling** [comment]
Dead code == code in comments,

**Comment by kimkulling** [comment]
There are some merge conflicts as well. I can take a look in the next days if you need any help!

**Comment by kimkulling** [comment]
Some thoughts from my side: would it be an option to use the Open3DGC-approach for the assimp-internal data structure as well? Our customers have a lot of big models to import and export.

**Comment by smalcom** [comment]
I read. I'll try to repair all.

**Comment by smalcom** [comment]
Need more time. Very busy at now.

**Comment by smalcom** [comment]
> dead code

Variable are set in CMakeLists.txt

```
FIND_PACKAGE(RT QUIET)
IF (RT_FOUND OR MSVC)
  ADD_DEFINITIONS( -DASSIMP_IMPORTER_GLTF_USE_OPEN3DGC )
```

Open3DGC codec has dependency in Linux - librt. I have thought that it is better not to do excess dependence for those who don't want to install this library.

> Hard to read, maybe 2 lines?

Agreed. Fixed.

> The common file header is missing, right?

Fixed.

> First finding: windows build is broken:

That is std::min/std::max used by author of the glTFAsset.xxx. I don't no why MSVC do not see it (((

**Comment by smalcom** [comment]
> This branch has conflicts that must be resolved
On my way

**Comment by smalcom** [comment]
Not my fault. Tested on fresh downloaded master. See attached files:
[unzip_cmake.txt](https://github.com/assimp/assimp/files/465270/unzip_cmake.txt)
[CMakeError.log.txt](https://github.com/assimp/assimp/files/465271/CMakeError.log.txt)
[CMakeOutput.log.txt](https://github.com/assimp/assimp/files/465272/CMakeOutput.log.txt)

**Comment by smalcom** [comment]
Only commit will start checks?

**Comment by smalcom** [comment]
Still same errors: CMakeLists.txt and std::min/max.
My log from linux:

```
smalcom [ /usr/src/build ]$ cmake ../assimp/
-- The C compiler identification is GNU 5.3.0
-- The CXX compiler identification is GNU 5.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Looking for ZLIB...
-- Checking for module 'zzip-zlib-config'
--   No package 'zzip-zlib-config' found
-- Found ZLIB: optimized;/usr/lib64/libz.so;debug;/usr/lib64/libz.so
-- Checking for module 'minizip'
--   No package 'minizip' found
-- Found rt: /usr/lib64/librt.so  
-- Enabled formats: 3DS AC ASE ASSBIN ASSXML B3D BVH COLLADA DXF CSM HMP IRRMESH IRR LWO LWS MD2 MD3 MD5 MDC MDL NFF NDO OFF OBJ OGRE OPENGEX PLY MS3D COB BLEND IFC XGL FBX Q3D Q3BSP RAW SIB SMD STL TERRAGEN 3D X GLTF 3MF
-- Disabled formats:
-- Found IL: /usr/lib64/libIL.so  
CMake Warning at CMakeLists.txt:358 (MESSAGE):
  Build of assimp_qt_viewer is disabled.  Unsatisfied dendencies: Qt5


-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Git: /usr/bin/git (found version "2.8.0") 
-- Configuring done
-- Generating done
-- Build files have been written to: /usr/src/build
```

**Comment by smalcom** [comment]
Successfully built with MSVS2013. Logs attached.

[build_log_assimp.7z.zip](https://github.com/assimp/assimp/files/465391/build_log_assimp.7z.zip)

**Comment by smalcom** [comment]
I've found that very fresh master is building successfully. I'm creating build system and trying to find mistake in my pull-request (((

**Comment by smalcom** [comment]
Finally! It's built ))

**Comment by kimkulling** [comment]
Sorry, you need to merge the latest master again. Rest looks good. Great work!

**Comment by smalcom** [comment]
*LOUD CRYING*

**Comment by smalcom** [comment]
Merging is done ))

**Comment by kimkulling** [comment]
Ha, it's in!

**Comment by kimkulling** [comment]
:-)

**Pull request merged**